### PR TITLE
MySQL compat: Implement %f in parseDateTime()

### LINKED ISSUE #48394: MySQL compatibility: format is not supported for fractional seconds
**Describe the unexpected behaviour**

ClickHouse complains about the usage of formatter `%f` which stands for fractional seconds (i.e. microseconds) in MySQL.

**How to reproduce**

* Which ClickHouse server version to use: the latest main branch version as of the 3th of April 2023 (`23.4.1.170`)
* Which interface to use, if matters: MySQL
* `CREATE TABLE` statements for all tables involved: use [cell towers](https://clickhouse.com/docs/en/getting-started/example-datasets/cell-towers) dataset, then query via MySQL protocol
* Queries to run that lead to unexpected result
```
SELECT CAST(CAST(DATE_FORMAT(created, '%Y-%m-%d %H:%i:%s') AS DATETIME) AS DATE) AS qt_a52bvfcv4c, COUNT(1) AS qt_up6h9rbv4c, area FROM cell_towers WHERE CAST(DATE_FORMAT(created, '%Y-%m-%d %H:%i:%s') AS DATETIME) BETWEEN STR_TO_DATE('2015-04-04T16:30:00.000000', '%Y-%m-%dT%H:%i:%s.%f') AND STR_TO_DATE('2017-04-04T16:00:00.000000', '%Y-%m-%dT%H:%i:%s.%f') GROUP BY qt_a52bvfcv4c, area;
```

**Expected behavior**

Date formatter works with fractional seconds 

**Error message and/or stacktrace**

```
SELECT CAST(CAST(DATE_FORMAT(created, '%Y-%m-%d %H:%i:%s') AS DATETIME) AS DATE) AS qt_a52bvfcv4c, COUNT(1) AS qt_up6h9rbv4c, area FROM cell_towers WHERE CAST(DATE_FORMAT(created, '%Y-%m-%d %H:%i:%s') AS DATETIME) BETWEEN STR_TO_DATE('2015-04-04T16:30:00.000000', '%Y-%m-%dT%H:%i:%s.%f') AND STR_TO_DATE('2017-04-04T16:00:00.000000', '%Y-%m-%dT%H:%i:%s.%f') GROUP BY qt_a52bvfcv4c, area; (stage: Complete)
2023.04.04 13:56:55.225668 [ 293 ] {mysql:41:57f0867a-4037-4b66-95c3-bd7b01a062c9} <Error> executeQuery: Code: 48. DB::Exception: format is not supported for fractional seconds: While processing SELECT CAST(CAST(DATE_FORMAT(created, '%Y-%m-%d %H:%i:%s'), 'DATETIME'), 'DATE') AS qt_a52bvfcv4c, count() AS qt_up6h9rbv4c, area FROM cell_towers WHERE (CAST(DATE_FORMAT(created, '%Y-%m-%d %H:%i:%s'), 'DATETIME') >= parseDateTimeOrNull('2015-04-04T16:30:00.000000', '%Y-%m-%dT%H:%i:%s.%f')) AND (CAST(DATE_FORMAT(created, '%Y-%m-%d %H:%i:%s'), 'DATETIME') <= parseDateTimeOrNull('2017-04-04T16:00:00.000000', '%Y-%m-%dT%H:%i:%s.%f')) GROUP BY qt_a52bvfcv4c, area LIMIT 150042. (NOT_IMPLEMENTED) (version 23.4.1.170 (official build)) (from 74.125.88.45:46416)
```

CC @rschu1ze

### PR DESCRIPTION
**Pull request by rschu1ze**
Fixes: #48394

This PR implements format string `%f` in `parseDateTime()`.

Note that `parseDateTime()` in ClickHouse produces a `DateTime` which does not support fractional seconds. This is different from MySQL in which `DateTime` does have fractional seconds. This gives two implementation options:
1. parse but ignore (truncate) the actual value
2. parse and round the result date's second fraction down up or up (depending on whether the value is > 500000 μs)

I found the first approach more consistent with the existing formatters (they don't do rounding).

Note: Function `formatDateTime()` is supposed to be the counterpart of `parseDateTime()`. Unfortunately, `formatDateTime(DateTime(...), '...%f...')` produces `0` which isn't consistent with this PR. I'll change that separately (need to think about backward compatibility ...). EDIT: --> #48422

CC: @taiyang-li 

### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
parseDateTime now understands format string '%f' (fractional seconds)

### CODE REVIEW

File: tests/queries/0_stateless/02668_parse_datetime.sql [hunk 1]
```diff
@@ -127,6 +127,13 @@ select parseDateTime('00/', '%s/', 'UTC') = toDateTime('1970-01-01 00:00:00', 'U
 select parseDateTime('60', '%s', 'UTC'); -- { serverError CANNOT_PARSE_DATETIME }
 select parseDateTime('-1', '%s', 'UTC'); -- { serverError CANNOT_PARSE_DATETIME }
 select parseDateTime('123456789', '%s', 'UTC'); -- { serverError CANNOT_PARSE_DATETIME }
+
+-- microsecond
+select parseDateTime('000000', '%f', 'UTC') = toDateTime('1970-01-01 00:00:00', 'UTC');
+select parseDateTime('456789', '%f', 'UTC') = toDateTime('1970-01-01 00:00:00', 'UTC');
```
Comment (antonio2368) [inline tests/queries/0_stateless/02668_parse_datetime.sql:133]: How do I know it doesn't return `1970-01-01 00:00:00` for all inputs 🤔

File: src/Functions/parseDateTime.cpp [hunk 2]
```diff
@@ -1035,6 +1035,37 @@ namespace
...
+                check_is_number(cur);
+                ++cur;
+                check_is_number(cur);
+                ++cur;
+                check_is_number(cur);
+                ++cur;
+                check_is_number(cur);
+                ++cur;
+                check_is_number(cur);
+                ++cur;
```
Comment (antonio2368) [inline src/Functions/parseDateTime.cpp:1065]: for loop

File: src/Functions/parseDateTime.cpp [hunk 3]
```diff
@@ -1035,6 +1035,37 @@ namespace
...
+                return cur;
```
Comment (antonio2368) [inline src/Functions/parseDateTime.cpp:1066]: where do you set date?

File: src/Functions/parseDateTime.cpp (hunk 2 above)
Comment (antonio2368) [reply src/Functions/parseDateTime.cpp:1065]: also what's up with `readNumber{2..4}`, why not extend it for 6?
And why can't it be a single function `template<size_t N>` with for loop inside?
I trust compiler should be able to do loop unwinding with constant loop size.

Comment (rschu1ze) [reply src/Functions/parseDateTime.cpp:1065]: for loop: done

`readNumber()` actually reads the number, whereas here, we want to read *over* the number (ignore it). But I'll add function `assertNumber()` which should make this more beautiful.

Comment (antonio2368) [reply src/Functions/parseDateTime.cpp:1065]: why does `mysqlSecond` then read the number? what's the difference?

File: src/Functions/parseDateTime.cpp (hunk 3 above)
Comment (rschu1ze) [reply src/Functions/parseDateTime.cpp:1066]: The output type (`DateTime`) supports no fractional seconds, therefore we just parse the input (and check at least that we read numbers) but ignore it (for details, see the PR message)

File: tests/queries/0_stateless/02668_parse_datetime.sql (hunk 1 above)
Comment (rschu1ze) [reply tests/queries/0_stateless/02668_parse_datetime.sql:133]: Good point, the test of `%f` in isolation look a bit weird. I'll extend the existing and more complex tests.

File: src/Functions/parseDateTime.cpp [hunk 4]
```diff
@@ -1035,6 +1053,13 @@ namespace
                 return cur;
             }
 
+            static Pos mysqlMicrosecond(Pos cur, Pos end, const String & fragment, DateTime & /*date*/)
+            {
+                for (size_t i = 0; i < 6; ++i)
+                    cur = assertNumber<NeedCheckSpace::Yes>(cur, end, fragment);
```
Comment (antonio2368) [inline src/Functions/parseDateTime.cpp:1059]: check space can be actually done before for loop 😅

**Review by antonio2368 — APPROVED** [review]

**Pull request merged**